### PR TITLE
feat: quantize Ideogram 4 with mflux-save

### PR DESCRIPTION
**Pull request by plz12345**
## Problem

`mflux-save --model ideogram-4-fp8 -q 4` produced a checkpoint that was still FP8, stamped with `quantization_level: 4`.

Every weight-bearing linear in Ideogram 4 is an `Fp8Linear` — both transformers and the Qwen3 text encoder. Two things kept `nn.quantize` away from all of them:

- `Fp8Linear` defined no `to_quantized`, and the default predicate is `hasattr(module, "to_quantized")`.
- Those three components were additionally declared `skip_quantization=True`.

So `-q` reached only the VAE. The model README already listed "quantization via `mflux-save`" as supported.

## Change

**`Fp8Linear.to_quantized`** dequantizes the FP8 codes through their per-row scale and requantizes to MLX affine. It has to serve both call orders in `WeightApplier`: quantizing a loaded FP8 checkpoint on the way to a save, and quantizing the skeleton of an already-quantized checkpoint *before* `update()` fills it in. The second path is why it returns a correctly shaped empty `QuantizedLinear` when the layer holds nothing yet.

**Dropped the three `skip_quantization=True` flags.**

**`_rebuild_q8_folded_layers` no longer assumes q8/group-64.** A 4-bit save packs twice as many weights into each `uint32`, so the hard-coded width sized the rebuilt layer wrongly and `update(strict=False)` then silently skipped every layer it had just replaced. Bits and group size are now derived from the stored shapes.

**Quantized embeddings are rebuilt as embeddings.** A quantized `nn.Embedding` stores exactly the same three tensors as a quantized linear (`weight`/`scales`/`biases`), so the payload cannot distinguish them — and the transformer's `embed_image_indicator` is an `nn.Embedding(2, emb_dim)`. Rebuilt as a linear, its lookup becomes a matmul against token indices and dies on the first forward:

```
[quantized_matmul] Last dimension of first input with shape (..., 271) does not match
the expanded quantized matrix (4608, 2) computed from shape (2,1152) with group_size=64,
bits=8 and transpose=true
```

This was latent before this PR: with the transformers skipped, no embedding ever reached that code carrying scales.

## Result

| Checkpoint | On disk | Save wall time | Peak RSS |
| --- | --- | --- | --- |
| FP8 (source) | 26 GB | — | — |
| `-q 8` | 26 GB | 34 s | 28.7 GB |
| `-q 4` | 14 GB | 32 s | 22.0 GB |

Quantizing loads the FP8 checkpoint once; MLX frees the FP8 weights as the graph evaluates, so peak resident stays under the sum of both copies and tracks the size of what is being written rather than the source. Measured on an M5 Pro / 64 GB, peak RSS in decimal GB.

### What it costs

Each quantized build is measured against an FP8 image generated in the same run on the same machine, as the fraction of subpixels differing at `rtol=0.1, atol=0`:

| Caption | noise floor | q8 vs FP8 | q4 vs FP8 |
| --- | --- | --- | --- |
| Photographic, 1024×1024 | 0.00% | 5.9% | 38.1% |
| Typographic poster, 768×576 | 0.00% | 29.2% | 65.6% |

Two FP8 runs at the same seed are bit-identical, so these are the cost of quantization and not run-to-run variance.

**They should be read as change, not damage.** q8 is visually indistinguishable from FP8. q4 is not — it renders the same prompt equally well but lands on a visibly different sample; compare the puffin's posture below. That is displacement, not degradation: diffusion is chaotic, so perturbing the weights moves a discrete decision — where a glyph lands, how a pose resolves — and the model settles into a different valid composition rather than a worse one. Typography holds at both levels.

The absolute figures track that reading. On the photographic caption q8 is 4.0 levels mean error (31.1 dB PSNR) while q4 is 22.1 levels (19.1 dB) — a real gap, and the one place where picking q4 means accepting a different image rather than the same one smaller.

**Photographic caption, 1024×1024**

| FP8 | q8 | q4 |
| --- | --- | --- |
| image | image | image |

**Typographic poster, 768×576**

| FP8 | q8 | q4 |
| --- | --- | --- |
| image | image| image |

## Tests

`tests/test_ideogram4_quantization.py` (fast, no weights) covers `to_quantized` preserving the forward at both widths, the pre-load skeleton path, bias being carried over and never invented, `nn.quantize` actually reaching `Fp8Linear`, the three components being quantizable, and the rebuild matching the stored bit width and keeping embeddings as embeddings.

`tests/image_generation/test_ideogram4_quantized_roundtrip.py` (slow, skipped unless the checkpoints are on disk) is the end-to-end proof. It loads each checkpoint exactly once and generates every case, so a full run costs three model loads rather than one per comparison.

```sh
mflux-save --model ideogram-4-fp8 -q 8 --path /path/to/ideogram-4-mflux-q8
mflux-save --model ideogram-4-fp8 -q 4 --path /path/to/ideogram-4-mflux-q4

export MFLUX_IDEOGRAM4_Q8_PATH=/path/to/ideogram-4-mflux-q8
export MFLUX_IDEOGRAM4_Q4_PATH=/path/to/ideogram-4-mflux-q4
pytest tests/image_generation/test_ideogram4_quantized_roundtrip.py -m slow -s
```

Two things that suite has to get right to mean anything, both of which caught me out while writing it:

**Both cases use JSON captions.** Ideogram 4's safety filter rejects plain prompts often enough that the main README's own `"A puffin standing on a cliff"` comes back blocked, and a blocked generation is returned as a flat grey field rather than an error. Two blocked outputs compare as near-identical, so a plain-prompt case reads as a passing test that proved nothing — my first run reported a reassuring 0.6% that was two blank images differing by dither. `test_fp8_baseline_was_actually_generated` now guards that permanently by asserting the baseline has structure (edge energy ~0.4 when blocked, 3.9–6.9 when real).

**Comparison is against a live FP8 baseline, not a checked-in PNG.** A stored reference also measures accumulated drift in the FP8 path and across platforms, and at `atol=0` a one-level difference on a dark subpixel counts the same as a catastrophic one.

Full fast suite green (555 passed), `ruff check` and `ruff format --check` clean.

## Notes

- Selective layer preservation — keeping small, sensitive layers at higher precision via the existing `quantization_predicate` hook — is deliberately out of scope. The existing uses of that hook in krea2 and the flux controlnet are correctness guards for non-64-divisible dims, which Ideogram 4 does not need (every layer is 64-divisible). A quality-motivated policy would need per-layer measurement to justify, and the visual result here gives it nothing to recover.
- `ModelSaver` writes no per-component `config.json`, and Ideogram 4 is the only model whose initializer reads one. Saved checkpoints reload because `_transformer_config` falls back to defaults that happen to match the official model exactly. That works today but leaves the artefact non-self-describing — worth a follow-up.
- The model README gained the `HF_TOKEN` half of the gated-weights requirement. It previously said access must be approved on the model card but never mentioned authenticating, which is the step people actually miss.

## Bug Fixes
- Fixed Ideogram 4 quantized checkpoint loading and reconstruction.
- Improved support for saved q4 and q8 checkpoints while preserving model behavior.
- Enabled reliable FP8-to-MLX quantization with configurable settings.

## Documentation
- Added guidance for gated model approval and authentication.
- Documented token methods, unauthenticated errors, quantization commands, local checkpoint creation, storage requirements, and performance observations.

## Tests
- Added coverage for FP8, q4, and q8 generation, checkpoint round trips, reproducibility, and output consistency.

<h3>Greptile Summary</h3>

The PR enables Ideogram 4 FP8 weights to be converted and round-tripped as MLX affine q4/q8 checkpoints.
- Adds loaded-weight and empty-skeleton conversion from `Fp8Linear` to `QuantizedLinear`.
- Reconstructs stored quantized layers using payload-derived bit width and group size while preserving embeddings.
- Enables quantization for both transformers and the Qwen3 text encoder.
- Adds focused unit tests, optional end-to-end generation checks, and usage documentation.

<h3>Confidence Score: 5/5</h3>

The PR appears safe to merge.

No blocking failure remains.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/mflux/models/ideogram4/model/ideogram4_transformer/fp8_linear.py | Adds FP8-to-MLX affine conversion for both loaded layers and pre-load skeletons. |
| src/mflux/models/ideogram4/ideogram4_initializer.py | Rebuilds saved q4/q8 payloads with inferred quantization parameters and preserves embedding semantics. |
| src/mflux/models/ideogram4/weights/ideogram4_weight_definition.py | Enables generic quantization for both Ideogram transformers and the text encoder. |
| tests/test_ideogram4_quantization.py | Covers FP8 conversion, skeleton loading, bias handling, component participation, bit-width reconstruction, and embeddings. |
| tests/image_generation/test_ideogram4_quantized_roundtrip.py | Adds optional end-to-end q4/q8 checkpoint generation comparisons against a live FP8 baseline. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Load Ideogram 4 checkpoint] --> B{Stored quantization metadata?}
    B -- No: FP8 source --> C[Apply FP8 weights]
    C --> D[nn.quantize calls Fp8Linear.to_quantized]
    D --> E[Dequantize FP8 using per-row scale]
    E --> F[Create MLX affine q4/q8 tensors]
    B -- Yes: saved q4/q8 --> G[Inspect stored weight, scales, and biases]
    G --> H[Infer bits and group size]
    H --> I[Rebuild QuantizedLinear or QuantizedEmbedding skeleton]
    I --> J[Apply stored quantized tensors]
    F --> K[Usable quantized model]
    J --> K
```
</details>

<sub>Reviews (3): Last reviewed commit: ["docs: measured peak RSS and wall time fo..."](https://github.com/mflux-community/mflux/commit/a50bb8b97923eff7b00ef31d8777fd62540a11a1) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=52078266)</sub>

### CODE REVIEW

**Comment by plz12345** [comment]
Note, I also added the pre-quantized models, using the code from this PR, to mflux-commuity on HuggingFace, and updated the HF org readme [matrix](https://huggingface.co/spaces/mflux-community/README/blob/main/README.md).

https://huggingface.co/mflux-community/ideogram-4-mflux-q4
https://huggingface.co/mflux-community/ideogram-4-mflux-q8

File: tests/image_generation/test_ideogram4_quantized_roundtrip.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        path.mkdir(parents=True, exist_ok=True)
+        return path
+    return tmp_path_factory.mktemp(f"ideogram4_{label}")
+
+
+@pytest.fixture(scope="module")
+def fp8_baselines(tmp_path_factory) -> dict[str, list[Path]]:
+    # Twice per case: the first is the baseline every quantized run is measured against,
+    # the second establishes the noise floor that makes those measurements mean anything.
+    return _generate_all(None, _destination(tmp_path_factory, "fp8"), repeats=2)
```
[the remark this answers is not in the document]
Comment (plz12345) [reply tests/image_generation/test_ideogram4_quantized_roundtrip.py:207]: Confirmed and fixed in 2279254.

Reproduced it before changing anything: with neither `MFLUX_IDEOGRAM4_Q8_PATH` nor `MFLUX_IDEOGRAM4_Q4_PATH` set, the module still loaded the FP8 model and generated four images, because `test_fp8_baseline_was_actually_generated` and `test_fp8_generation_is_reproducible` take the `fp8_baselines` fixture without going through `_require_checkpoint`. On a machine that has never pulled the gated weights that is a 26 GB download, and the module's own docstring promises it skips.

Guarded the fixture rather than the individual tests, so every consumer is covered, and reused the same helper in `quantized_outputs` so there is one definition of "which checkpoints are actually present". With nothing configured the module now skips in ~1 s; with both configured the comparisons still run.

**Review by fxd0h — APPROVED** [review]
Reproduced the whole path on an M5 Max, source checkpoint on an external disk and the save on the internal one: `-q 4` wrote 14 GB in 43 s, matching your figure. Loading it back takes 1.5 s and the text encoder comes up as 252 QuantizedLinear plus one QuantizedEmbedding, all reporting bits=4 group_size=64, so the inference recovers what was written and the embedding stays an embedding. The puffin case at V4_TURBO_12 generates in 68 s with edge energy 3.74, and it is a clean puffin rather than a filtered blank.

Also checked `to_quantized` on a synthetic fp8 layer across more widths than this needs, comparing the quantized forward against the fp8 one:

| bits | group | packed weight | rel err vs fp8 |
|---|---|---|---|
| 8 | 64 | (512, 128) | 0.008 |
| 6 | 64 | (512, 96) | 0.023 |
| 4 | 64 | (512, 64) | 0.091 |
| 3 | 64 | (512, 48) | 0.192 |

Packed widths are exact at every level, group 32 behaves the same, and the skeleton path returns a correctly shaped empty layer rather than raising. Fast suite green here.

One note for later. Bits and group size are not separately determined by the stored shapes, only their product is: packed times 32 over num_groups equals group_size times bits, so (64, 4) and (32, 8) describe the same tensors. `_infer_quantization` breaks the tie with the layer's own inner dimension, which is right, and the `num_groups * 64` fallback is the one path that cannot. Not reachable today, since every embedding here carries a weight to read the dimension from. #517 adds the same computation generically in `WeightApplier` with a plausibility check on the result, so worth consolidating on one of them later. No conflict between the two now.
